apis/kueue/v1alpha2/clusterqueue_types.go

```diff
@@ -200,7 +201,7 @@ type Resource struct {
 type Flavor struct {
 	// name is a reference to the resourceFlavor that defines this flavor.
 	// +kubebuilder:default=default
-	Name ResourceFlavorReference `json:"name"`
+	Name ResourceFlavorReference `json:"name,omitempty"`
```

Comment:
To work `kubebuilder:default=default`, we probably need to add `omitempty` as a yaml marker in the name field.

Reply:
If we do not add `omitempty`, all empty names will be rejected by the CRD validation.

Reply:
> To work kubebuilder:default=default, we probably need to add omitempty as a yaml marker in the name field.

did you test that this is indeed the case?

Reply:
Sure. you can reproduce using my PR branch if the name is defaulting or not.

Reply:
1. deploy kueue using my branch.

```shell
kubectl apply -k github.com/tenzen-y/kueue/config/default?ref=fix-comments-for-flavor
```

2. deploy CR including the empty flavor name

```shell
cat <<EOF| kubectl apply -f -
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - quota:
        min: 9
  - name: "memory"
    flavors:
    - quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
  clusterQueue: cluster-total
EOF
```

3. check result

```shell
kubectl get clusterqueues.kueue.x-k8s.io cluster-total -oyaml
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"kueue.x-k8s.io/v1alpha2","kind":"ClusterQueue","metadata":{"annotations":{},"name":"cluster-total"},"spec":{"namespaceSelector":{},"resources":[{"flavors":[{"quota":{"min":9}}],"name":"cpu"},{"flavors":[{"quota":{"min":"36Gi"}}],"name":"memory"}]}}
  creationTimestamp: "2022-09-12T18:03:08Z"
  finalizers:
  - kueue.k8s.io/resource-in-use
  generation: 1
  name: cluster-total
  resourceVersion: "673"
  uid: 77f3bd7c-3137-4bc8-a917-408b1a085cb3
spec:
  namespaceSelector: {}
  queueingStrategy: BestEffortFIFO
  resources:
  - flavors:
    - name: default
      quota:
        min: "9"
    name: cpu
  - flavors:
    - name: default
      quota:
        min: 36Gi
    name: memory
status:
  admittedWorkloads: 0
  pendingWorkloads: 0
  usedResources:
    cpu:
      default:
        total: "0"
    memory:
      default:
        total: "0"
```

Reply:
I faced validation errors if I used v0.2.1 in the following:

```shell
error validating "STDIN": error validating data: [ValidationError(ClusterQueue.spec.resources[0].flavors[0]): missing required field "name" in io.x-k8s.kueue.v1alpha2.ClusterQueue.spec.resources.flavors, ValidationError(ClusterQueue.spec.resources[1].flavors[0]): missing required field "name" in io.x-k8s.kueue.v1alpha2.ClusterQueue.spec.resources.flavors]; if you choose to ignore these errors, turn validation off with --validate=false
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "mresourceflavor.kb.io": failed to call webhook: Post "https://kueue-webhook-service.kueue-system.svc:443/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor?timeout=10s": dial tcp 10.96.169.215:443: connect: connection refused
```

Reply:
Is it failing consistently? It looks like the webhook was unreachable, which is not outside the realm of posibilities.

Reply:
Unless the webhook/controller crashed? did you look at the logs?

Reply:
Sorry, I sent the wrong log.
I show the correct logs in below:

```shell
$ cat <<EOF| kubectl apply -f -
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - quota:
        min: 9
  - name: "memory"
    flavors:
    - quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
  clusterQueue: cluster-total
EOF
resourceflavor.kueue.x-k8s.io/default created
error: error validating "STDIN": error validating data: [ValidationError(ClusterQueue.spec.resources[0].flavors[0]): missing required field "name" in io.x-k8s.kueue.v1alpha2.ClusterQueue.spec.resources.flavors, ValidationError(ClusterQueue.spec.resources[1].flavors[0]): missing required field "name" in io.x-k8s.kueue.v1alpha2.ClusterQueue.spec.resources.flavors]; if you choose to ignore these errors, turn validation off with --validate=false
```

> Is it failing consistently? It looks like the webhook was unreachable, which is not outside the realm of possibilities.

I think that the name field is validated by [OpenAPI CRD validation](https://book.kubebuilder.io/reference/markers/crd-validation.html#crd-validation) specified in [this](https://github.com/kubernetes-sigs/kueue/blob/6e403b517b8a3bfcc71b7ded10be9ca9d8634c5b/config/components/crd/bases/kueue.x-k8s.io_clusterqueues.yaml#L224-L226), so the undefined name field is rejected before the requested data (ClusterQueue object) reaches validating webhooks.

Actually, we can see validation errors message from kueue-validating-webhook  if we send CluserQueue with the empty name field to kube-apiserver in the below.

```shell
cat <<EOF| kubectl apply -f -
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - name: ""
      quota:
        min: 9
  - name: "memory"
    flavors:
    - name: ""
      quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
  clusterQueue: cluster-total
EOF
resourceflavor.kueue.x-k8s.io/default unchanged
localqueue.kueue.x-k8s.io/main created
Error from server ([spec.resources[0].flavors[0].name: Invalid value: "": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.resources[1].flavors[0].name: Invalid value: "": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]): error when creating "STDIN": admission webhook "vclusterqueue.kb.io" denied the request: [spec.resources[0].flavors[0].name: Invalid value: "": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.resources[1].flavors[0].name: Invalid value: "": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]
```

> Unless the webhook/controller crashed? did you look at the logs?

Just in case, I show the kueue-controller logs:

```shell
$ kubectl get pods -n kueue-system kueue-controller-manager-7b9946586c-s2qdt 
NAME                                        READY   STATUS    RESTARTS   AGE
kueue-controller-manager-7b9946586c-s2qdt   2/2     Running   0          10s
```

<details>
<summary>full logs</summary>

```shell
{"level":"info","ts":"2022-09-13T03:28:19.907230297Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Successfully loaded config file","config":"apiVersion: config.kueue.x-k8s.io/v1alpha1\nenableInternalCertManagement: true\nhealth:\n  healthProbeBindAddress: :8081\nkind: Configuration\nleaderElection:\n  leaderElect: true\n  leaseDuration: 0s\n  renewDeadline: 0s\n  resourceLock: \"\"\n  resourceName: c1f6bfd2.kueue.x-k8s.io\n  resourceNamespace: \"\"\n  retryPeriod: 0s\nmanageJobsWithoutQueueName: false\nmetrics:\n  bindAddress: :8080\nwebhook:\n  port: 9443\n"}
{"level":"info","ts":"2022-09-13T03:28:20.204243548Z","logger":"controller-runtime.metrics","caller":"logr@v1.2.3/logr.go:261","msg":"Metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":"2022-09-13T03:28:20.283492298Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Probe endpoints are configured on healthz and readyz"}
{"level":"info","ts":"2022-09-13T03:28:20.284264048Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Waiting for certificate generation to complete"}
{"level":"info","ts":"2022-09-13T03:28:20.285830798Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Starting manager"}
{"level":"info","ts":"2022-09-13T03:28:20.290002089Z","caller":"manager/internal.go:362","msg":"Starting server","kind":"health probe","addr":"[::]:8081"}
{"level":"info","ts":"2022-09-13T03:28:20.289609381Z","caller":"manager/internal.go:362","msg":"Starting server","path":"/metrics","kind":"metrics","addr":"[::]:8080"}
I0913 03:28:20.399512       1 leaderelection.go:248] attempting to acquire leader lease kueue-system/c1f6bfd2.kueue.x-k8s.io...
I0913 03:28:36.621897       1 leaderelection.go:258] successfully acquired lease kueue-system/c1f6bfd2.kueue.x-k8s.io
{"level":"debug","ts":"2022-09-13T03:28:36.622054472Z","logger":"events","caller":"recorder/recorder.go:103","msg":"Normal","object":{"kind":"Lease","namespace":"kueue-system","name":"c1f6bfd2.kueue.x-k8s.io","uid":"9343edaa-d219-43b6-af1d-2d0423324fde","apiVersion":"coordination.k8s.io/v1","resourceVersion":"12170"},"reason":"LeaderElection","message":"kueue-controller-manager-7b9946586c-s2qdt_b13e8b8f-c3ee-4674-bfd6-a12025876d7f became leader"}
{"level":"info","ts":"2022-09-13T03:28:36.624885097Z","logger":"cert-rotation","caller":"logr@v1.2.3/logr.go:261","msg":"starting cert rotator controller"}
{"level":"info","ts":"2022-09-13T03:28:36.67937293Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"cert-rotator","source":"&{{%!s(*v1.Secret=&{{ } {      0 {{0 0 <nil>}} <nil> <nil> map[] map[] [] []  []} <nil> map[] map[] }) %!s(*cache.informerCache=&{0xc0004deda0}) %!s(chan error=<nil>) %!s(func()=<nil>)}}"}
{"level":"info","ts":"2022-09-13T03:28:36.680876222Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"cert-rotator","source":"&{{%!s(*unstructured.Unstructured=&{map[apiVersion:admissionregistration.k8s.io/v1 kind:ValidatingWebhookConfiguration]}) %!s(*cache.informerCache=&{0xc0004deda0}) %!s(chan error=<nil>) %!s(func()=<nil>)}}"}
{"level":"info","ts":"2022-09-13T03:28:36.681749264Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"cert-rotator","source":"&{{%!s(*unstructured.Unstructured=&{map[apiVersion:admissionregistration.k8s.io/v1 kind:MutatingWebhookConfiguration]}) %!s(*cache.informerCache=&{0xc0004deda0}) %!s(chan error=<nil>) %!s(func()=<nil>)}}"}
{"level":"info","ts":"2022-09-13T03:28:36.682268347Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"cert-rotator"}
{"level":"info","ts":"2022-09-13T03:28:36.788624305Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"cert-rotator","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:36.789610805Z","logger":"cert-rotation","caller":"logr@v1.2.3/logr.go:261","msg":"no cert refresh needed"}
{"level":"info","ts":"2022-09-13T03:28:36.792636764Z","logger":"cert-rotation","caller":"rotator/rotator.go:685","msg":"Ensuring CA cert","name":"kueue-validating-webhook-configuration","gvk":"admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration","name":"kueue-validating-webhook-configuration","gvk":"admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration"}
{"level":"info","ts":"2022-09-13T03:28:36.87411693Z","logger":"cert-rotation","caller":"logr@v1.2.3/logr.go:261","msg":"certs are ready in /tmp/k8s-webhook-server/serving-certs"}
{"level":"info","ts":"2022-09-13T03:28:36.885083222Z","logger":"cert-rotation","caller":"rotator/rotator.go:685","msg":"Ensuring CA cert","name":"kueue-mutating-webhook-configuration","gvk":"admissionregistration.k8s.io/v1, Kind=MutatingWebhookConfiguration","name":"kueue-mutating-webhook-configuration","gvk":"admissionregistration.k8s.io/v1, Kind=MutatingWebhookConfiguration"}
{"level":"info","ts":"2022-09-13T03:28:38.238249917Z","logger":"cert-rotation","caller":"logr@v1.2.3/logr.go:261","msg":"CA certs are injected to webhooks"}
{"level":"info","ts":"2022-09-13T03:28:38.238908667Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Certs ready"}
{"level":"info","ts":"2022-09-13T03:28:38.242619542Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"resourceflavor","controllerGroup":"kueue.x-k8s.io","controllerKind":"ResourceFlavor"}
{"level":"info","ts":"2022-09-13T03:28:38.242863542Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"resourceflavor","controllerGroup":"kueue.x-k8s.io","controllerKind":"ResourceFlavor","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:38.243479417Z","caller":"controller/controller.go:150","msg":"Starting EventSource","controller":"resourceflavor","controllerGroup":"kueue.x-k8s.io","controllerKind":"ResourceFlavor","source":"kind source: *v1alpha2.ResourceFlavor"}
{"level":"info","ts":"2022-09-13T03:28:38.244439125Z","caller":"controller/controller.go:150","msg":"Starting EventSource","controller":"resourceflavor","controllerGroup":"kueue.x-k8s.io","controllerKind":"ResourceFlavor","source":"channel source: 0xc0004082d0"}
{"level":"info","ts":"2022-09-13T03:28:38.248351542Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"localqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"LocalQueue","source":"kind source: *v1alpha2.LocalQueue"}
{"level":"info","ts":"2022-09-13T03:28:38.248421Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"ClusterQueue","source":"kind source: *v1alpha2.ClusterQueue"}
{"level":"info","ts":"2022-09-13T03:28:38.248721917Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"ClusterQueue","source":"kind source: *v1.Namespace"}
{"level":"info","ts":"2022-09-13T03:28:38.250424667Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"localqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"LocalQueue","source":"channel source: 0xc000153310"}
{"level":"info","ts":"2022-09-13T03:28:38.250489542Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"ClusterQueue","source":"channel source: 0xc0001533b0"}
{"level":"info","ts":"2022-09-13T03:28:38.251221292Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"ClusterQueue"}
{"level":"info","ts":"2022-09-13T03:28:38.251634958Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"workload","controllerGroup":"kueue.x-k8s.io","controllerKind":"Workload","source":"kind source: *v1alpha2.Workload"}
{"level":"info","ts":"2022-09-13T03:28:38.251950375Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"workload","controllerGroup":"kueue.x-k8s.io","controllerKind":"Workload"}
{"level":"info","ts":"2022-09-13T03:28:38.253154958Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"job","controllerGroup":"batch","controllerKind":"Job","source":"kind source: *v1.Job"}
{"level":"info","ts":"2022-09-13T03:28:38.253444333Z","caller":"controller/controller.go:185","msg":"Starting EventSource","controller":"job","controllerGroup":"batch","controllerKind":"Job","source":"kind source: *v1alpha2.Workload"}
{"level":"info","ts":"2022-09-13T03:28:38.253529667Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"job","controllerGroup":"batch","controllerKind":"Job"}
{"level":"info","ts":"2022-09-13T03:28:38.251320083Z","caller":"controller/controller.go:193","msg":"Starting Controller","controller":"localqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"LocalQueue"}
{"level":"info","ts":"2022-09-13T03:28:38.274770083Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a mutating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=Workload","path":"/mutate-kueue-x-k8s-io-v1alpha2-workload"}
{"level":"info","ts":"2022-09-13T03:28:38.275249708Z","logger":"controller-runtime.webhook.webhooks","caller":"webhook/server.go:216","msg":"Starting webhook server"}
{"level":"info","ts":"2022-09-13T03:28:38.279289958Z","logger":"controller-runtime.certwatcher","caller":"logr@v1.2.3/logr.go:261","msg":"Updated current TLS certificate"}
{"level":"info","ts":"2022-09-13T03:28:38.281561708Z","logger":"controller-runtime.webhook","caller":"logr@v1.2.3/logr.go:261","msg":"Serving webhook server","host":"","port":9443}
{"level":"info","ts":"2022-09-13T03:28:38.284816333Z","logger":"controller-runtime.certwatcher","caller":"logr@v1.2.3/logr.go:261","msg":"Starting certificate watcher"}
{"level":"info","ts":"2022-09-13T03:28:38.286884167Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/mutate-kueue-x-k8s-io-v1alpha2-workload"}
{"level":"info","ts":"2022-09-13T03:28:38.288342208Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a validating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=Workload","path":"/validate-kueue-x-k8s-io-v1alpha2-workload"}
{"level":"info","ts":"2022-09-13T03:28:38.288774458Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/validate-kueue-x-k8s-io-v1alpha2-workload"}
{"level":"info","ts":"2022-09-13T03:28:38.289757833Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a mutating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=ResourceFlavor","path":"/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor"}
{"level":"info","ts":"2022-09-13T03:28:38.290165542Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor"}
{"level":"info","ts":"2022-09-13T03:28:38.290479958Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a validating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=ResourceFlavor","path":"/validate-kueue-x-k8s-io-v1alpha2-resourceflavor"}
{"level":"info","ts":"2022-09-13T03:28:38.290778625Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/validate-kueue-x-k8s-io-v1alpha2-resourceflavor"}
{"level":"info","ts":"2022-09-13T03:28:38.291265667Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a mutating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=ClusterQueue","path":"/mutate-kueue-x-k8s-io-v1alpha2-clusterqueue"}
{"level":"info","ts":"2022-09-13T03:28:38.291764958Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/mutate-kueue-x-k8s-io-v1alpha2-clusterqueue"}
{"level":"info","ts":"2022-09-13T03:28:38.351412667Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"localqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"LocalQueue","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:38.352871792Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"clusterqueue","controllerGroup":"kueue.x-k8s.io","controllerKind":"ClusterQueue","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:38.353165709Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"workload","controllerGroup":"kueue.x-k8s.io","controllerKind":"Workload","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:38.373155667Z","caller":"controller/controller.go:227","msg":"Starting workers","controller":"job","controllerGroup":"batch","controllerKind":"Job","worker count":1}
{"level":"info","ts":"2022-09-13T03:28:38.373388125Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a validating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=ClusterQueue","path":"/validate-kueue-x-k8s-io-v1alpha2-clusterqueue"}
{"level":"info","ts":"2022-09-13T03:28:38.37371725Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/validate-kueue-x-k8s-io-v1alpha2-clusterqueue"}
{"level":"info","ts":"2022-09-13T03:28:38.374389917Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"skip registering a mutating webhook, object does not implement admission.Defaulter or WithDefaulter wasn't called","GVK":"kueue.x-k8s.io/v1alpha2, Kind=LocalQueue"}
{"level":"info","ts":"2022-09-13T03:28:38.374566459Z","logger":"controller-runtime.builder","caller":"logr@v1.2.3/logr.go:261","msg":"Registering a validating webhook","GVK":"kueue.x-k8s.io/v1alpha2, Kind=LocalQueue","path":"/validate-kueue-x-k8s-io-v1alpha2-localqueue"}
{"level":"info","ts":"2022-09-13T03:28:38.374825417Z","logger":"controller-runtime.webhook","caller":"webhook/server.go:148","msg":"Registering webhook","path":"/validate-kueue-x-k8s-io-v1alpha2-localqueue"}
{"level":"debug","ts":"2022-09-13T03:28:46.831692921Z","logger":"controller-runtime.webhook.webhooks","caller":"admission/http.go:96","msg":"received request","webhook":"/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor","UID":"d4c5cb41-d6af-424e-a6b5-6eaee0992c36","kind":"kueue.x-k8s.io/v1alpha2, Kind=ResourceFlavor","resource":{"group":"kueue.x-k8s.io","version":"v1alpha2","resource":"resourceflavors"}}
{"level":"debug","ts":"2022-09-13T03:28:46.836001754Z","logger":"controller-runtime.webhook.webhooks","caller":"admission/http.go:143","msg":"wrote response","webhook":"/mutate-kueue-x-k8s-io-v1alpha2-resourceflavor","code":200,"reason":"","UID":"d4c5cb41-d6af-424e-a6b5-6eaee0992c36","allowed":true}
{"level":"debug","ts":"2022-09-13T03:28:46.850457088Z","logger":"controller-runtime.webhook.webhooks","caller":"admission/http.go:96","msg":"received request","webhook":"/validate-kueue-x-k8s-io-v1alpha2-resourceflavor","UID":"c26540a0-3905-4900-a705-d7553176891e","kind":"kueue.x-k8s.io/v1alpha2, Kind=ResourceFlavor","resource":{"group":"kueue.x-k8s.io","version":"v1alpha2","resource":"resourceflavors"}}
{"level":"debug","ts":"2022-09-13T03:28:46.851153004Z","logger":"controller-runtime.webhook.webhooks","caller":"admission/http.go:143","msg":"wrote response","webhook":"/validate-kueue-x-k8s-io-v1alpha2-resourceflavor","code":200,"reason":"","UID":"c26540a0-3905-4900-a705-d7553176891e","allowed":true}
{"level":"Level(-2)","ts":"2022-09-13T03:28:46.856292838Z","logger":"resourceflavor-reconciler","caller":"core/resourceflavor_controller.go:107","msg":"ResourceFlavor create event","resourceFlavor":{"name":"default"}}
{"level":"Level(-2)","ts":"2022-09-13T03:28:46.857807046Z","caller":"core/resourceflavor_controller.go:69","msg":"Reconciling ResourceFlavor","controller":"resourceflavor","controllerGroup":"kueue.x-k8s.io","controllerKind":"ResourceFlavor","resourceFlavor":{"name":"default"},"namespace":"","name":"default","reconcileID":"0f09782e-f7f9-4189-96c0-6767da37697a","resourceFlavor":{"name":"default"}}
```
</details>

Reply:
ok, now that we have a validation and defaulting webhook it makes sense to proceed with this change.

In the past, we didn't have them, so it was better to fail on empty string.

Reply:
Btw, what happens if you send two flavors for the same resource type with empty name (in your branch)?

Reply:
Is this what you mean?

- empty name:

```shell
$ cat <<EOF| kubectl apply -f -
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - name: ""
      quota:
        min: 9
    - name: ""
      quota:
        min: 6
  - name: "memory"
    flavors:
    - name: ""
      quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
EOFlusterQueue: cluster-total
resourceflavor.kueue.x-k8s.io/default created
localqueue.kueue.x-k8s.io/main created
The ClusterQueue "cluster-total" is invalid: spec.resources[0].flavors[1]: Duplicate value: map[string]interface {}{"name":""}
```

- undefined name:

```shell
cat <<EOF| kubectl apply -f -
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ResourceFlavor
metadata:
  name: default
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: ClusterQueue
metadata:
  name: cluster-total
spec:
  namespaceSelector: {}
  resources:
  - name: "cpu"
    flavors:
    - quota:
        min: 9
    - quota:
        min: 6
  - name: "memory"
    flavors:
    - quota:
        min: 36Gi
---
apiVersion: kueue.x-k8s.io/v1alpha2
kind: LocalQueue
metadata:
  namespace: default
  name: main
spec:
  clusterQueue: cluster-total
EOF
resourceflavor.kueue.x-k8s.io/default created
localqueue.kueue.x-k8s.io/main created
The ClusterQueue "cluster-total" is invalid: spec.resources[0].flavors[1]: Duplicate value: map[string]interface {}{"name":"default"}
```

Reply:
> For this PR (about fixing typos), just revert this line.

That makes sense.

Reply:
Reverted this line.